### ISSUE TITLE
Avoid String.Format when interpolated strings are statically known

### ISSUE THREAD

**Issue by danmosemsft** [dotnet/roslyn#22344]
**Version Used**: 
2.0.2-vspre-006949

**Steps to Reproduce**:
using System;
public class C {
    public void M(string arg) {
       var knownStatically = "boz";
       Console.WriteLine($"{nameof(arg)} baz {knownStatically}");
    }
}
[live repro](
https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxASwDZICYgNQA+AAgAwAERAjANwCwAUEQMwUBMZAwmQN4Nn8UWRACxkAsgAoq5CAgDmASh58BAgG6yyAawB2AewDuOgMowIMNAGMIGDAE8yAXjIAiAGZ69LuvVUCqAJwSACQu3DoQYHB6bhKyigC+ZMAQAF48uoYmZhbWtnYJLgo+qgkMCUA==)

**Expected Behavior**:
IL contains
```
        IL_0007: ldstr "arg baz boz"
```

**Actual Behavior**:
IL contains 
```
        IL_0007: ldstr "{0} baz {1}"
        IL_000c: ldstr "arg"
        IL_0011: ldloc.0
        IL_0012: call string [mscorlib]System.String::Format(string, object, object)
```

**Motivation**
String.Format is not a cheap operation. This can be hit in cases like asserts, eg.,
``` c#
Debug.Assert(Capacity == Length, $"{nameof(ExpandByABlock)} should only be called when there is no space left.");
```
The above causes a `String.Format` which news up a `StringBuilder` even if the condition is true. Similar patterns can exist in retail builds.

/cc @stephentoub

**Comment by sharwell** [comment]
This appears to be a duplicate of #6122 and/or #4678.

**Comment by tannergooding** [comment]
I thought it was determine by @gafter this couldn't be done due to the possibility of custom runtimes and/or custom cultures.

That being said, this is one of the many optimizations that I wish we had a "please make this assumption and optimize appropriately" switch for.

I think last time it was brought up, it was mentioned that a general-purpose, post-compilation, IL optimization tool would be a better investment.

**Comment by stephentoub** [comment]
In this case, the const in question is a string; culture would not affect that (string.ToString() is string, regardless of culture).  And a custom runtime?  We're seriously concerned about a runtime that implements string.Format to have different results than does .NET Framework, .NET Core, and Mono?

**Comment by tannergooding** [comment]
@stephentoub, it isn't calling `string.ToString()` though. It is calling `string.ToString(IFormattable)` which can be affected by Culture (however, on netfx, netcore, and mono all just behaves the same as `string.ToString()`).

> We're seriously concerned about a runtime that implements string.Format to have different results than does .NET Framework, .NET Core, and Mono?

That's what has come up on several proposals and issues. Due to the existence of custom runtimes, there are several optimizations and or assumptions that cannot be made.

**Comment by tannergooding** [comment]
I don't necessarily agree with it, its just what has been stated time and time again.

**Comment by stephentoub** [comment]
> it isn't calling string.ToString() though. It is calling string.ToString(IFormattable) which can be affected by Culture

No, that's my point... there is no formatting be done as part of ToString, regardless of overload... ToString on a string just returns that string, regardless of culture.

> Due to the existence of custom runtimes, there are several optimizations and or assumptions that cannot be made.

And especially in this case, I think that's irrelevant.  We're shooting ourselves in the feet by taking such a stance.

I realize you're just the messenger...

**Comment by danmosemsft** [comment]
This means we should scan for a new antipattern in code reviews, which I wasn't previously aware of. I would not have expected lines like this, even in release to allocate a StringBuilder even when the condition is false.
```
Trace.WriteIf(someFlag, $"{nameof(SomeFunction)} should only be called when {nameof(someFlag)} is true.");
```

**Comment by stephentoub** [comment]
> This means we should scan for a new antipattern in code reviews

I consider any use of string interpolation forbidden in corefx on any code paths where we might care about perf.  They're fine in asserts, in exceptions, etc., but this issue aside, they're far from cheap.

> I would not have expected lines like this, even in release to allocate a StringBuilder

It's using string.Format.  And string.Format uses StringBuilderCache, which will allocate a StringBuilder if it doesn't already have one.

**Comment by CyrusNajmabadi** [comment]
@333fred is this an issue anymore now that we have interpolation handlers?

**Comment by 333fred** [comment]
This is still asking for something more, but I don't know if we'd ever actually do this. @jaredpar ?

**Comment by jaredpar** [comment]
I doubt this would ever arise to the priority that we'd do it but we'd accept a PR that did it. That's why it's in Backlog

**Comment by svick** [comment]
@333fred 

> This is still asking for something more, but I don't know if we'd ever actually do this.

[The example in the original post now compiles to `string.Concat("arg baz ", knownStatically)`](https://sharplab.io/#v2:D4AQTAjAsAUCDMACciDCiDetE+UkALIgLIAUIEADIgIYBOA5gJSba64Bu9iA1gHYB7AO58AygBca4gJYBjGgBsFAT0QBeRACIAZgIGaA3G3bIIATlIASTRj40AtgFMB20vWYBfRACMaAL0x+YTFJGXklZQ9NJiMYdg9YDyA=) or [to just `"arg baz foo"` if the variable is turned into a `const`](https://sharplab.io/#v2:D4AQTAjAsAUCDMACciDCiDetE+UkALIgLIAUIEADIgIYBOA5gJSba64DGA9gHYDOAF2RVEAax5cA7jwDKAmgICWHGgBtVAT0QBeRACIAZly56A3G3bCAnKQAkejDxoBbAKZcDpeswC+iAEY0AF6Y4lKy8koq6ho+ekzmMOw+sD5AA===).

What more is to be done here?
